File: apps/mobile/ios/Sources/Flagship/Recovery.swift
import Foundation
import CryptoKit
import AuthenticationServices

/// WebAuthn-PRF cloud recovery — J.3/J.4 in the architecture spec.
///
/// Setup (on the user's primary phone, after onboarding):
///   1. Register a platform passkey for `flagshipserver.com` via
///      ASAuthorizationPlatformPublicKeyCredentialRegistrationRequest.
///   2. Run a PRF-extension assertion to derive a 32-byte secret from
///      the passkey + a fixed salt.
///   3. AES-GCM encrypt the UMK seed under the PRF-derived key.
///   4. Upload a SIGNED {credentialId, wrappedUmk, issuedAt} envelope to
///      flagshipserver.com (POST /api/recovery). wrappedUmk is the single
///      self-contained AES-GCM blob (nonce‖ct‖tag); see RecoveryUpload for
///      the canonical bytes the IRK signs.
///
/// Recovery (on a fresh phone — old phone lost):
///   1. PRF assertion against the registered credential.
///   2. Fetch envelope from flagshipserver.com keyed by credentialID.
///   3. AES-GCM decrypt → recover the original UMK seed.
///   4. Inject into Keystore so derived keys (BAK/IRK/SWK) match.
///
/// The PRF extension (CTAP2 hmac-secret) isn't surfaced by Apple's
/// AuthenticationServices framework prior to iOS 18, and even on 18+
/// it has caveats (requires .platformProvider with explicit PRF
/// support). This module models the round-trip API so the UI flow
/// works end-to-end against the mock surface; the live impl needs an
/// ASAuthorizationController delegate to do the actual ceremony.
public struct Recovery {

    public enum RecoveryError: Error, LocalizedError {
        case prfUnavailable
        case envelopeMissing
        case decryptionFailed(String)
        case keystoreError(String)

        public var errorDescription: String? {
            switch self {
            case .prfUnavailable:        return "Your authenticator doesn't support PRF; pick another."
            case .envelopeMissing:       return "No recovery envelope is registered for that credential."
            case .decryptionFailed(let m): return "Couldn't decrypt the UMK: \(m)"
            case .keystoreError(let m):    return "Keystore error: \(m)"
            }
        }
    }

    /// Wrap a UMK seed under a PRF-derived key. Returns a SINGLE
    /// self-contained base64 blob — the AES-GCM `.combined` representation
    /// (nonce‖ciphertext‖tag) — ready to ship as the Worker's `wrappedUmk`
    /// field. The nonce lives inside the blob; there is no separate nonce
    /// field. The Worker SHA-256s the decoded bytes of exactly this blob.
    public static func wrap(
        umkSeed: SymmetricKey,
        prfSecret: Data
    ) throws -> String {
        let key = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: prfSecret),
            salt: Data("flagship/recovery-wrap/v1".utf8),
            info: Data(),
            outputByteCount: 32
        )
        let umkBytes = umkSeed.withUnsafeBytes { Data($0) }
        let sealed = try AES.GCM.seal(umkBytes, using: key)
        guard let combined = sealed.combined else {
            throw RecoveryError.decryptionFailed("no combined ciphertext")
        }
        return combined.base64EncodedString()
    }

    /// Reverse of `wrap`. Decrypts the single self-contained combined blob
    /// (nonce‖ciphertext‖tag, base64) into a UMK seed.
    public static func unwrap(
        wrappedUmkBase64: String,
        prfSecret: Data
    ) throws -> SymmetricKey {
        guard let combined = Data(base64Encoded: wrappedUmkBase64) else {
            throw RecoveryError.decryptionFailed("base64 decode")
        }
        let key = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: prfSecret),
            salt: Data("flagship/recovery-wrap/v1".utf8),
            info: Data(),
            outputByteCount: 32
        )
        do {
            let box = try AES.GCM.SealedBox(combined: combined)
            let plaintext = try AES.GCM.open(box, using: key)
            return SymmetricKey(data: plaintext)
        } catch {
            throw RecoveryError.decryptionFailed(String(describing: error))
        }
    }

    /// Fixed PRF input salt — `hmac-secret` is keyed by salt, so the
    /// daemon side and recovery side must agree. Mirrors the constant
    /// in `packages/server-daemon/src/recovery/prf.ts`.
    public static let prfSalt: Data = Data("flagship/recovery/v1".utf8)
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/keystore/Recovery.kt
// Cloud recovery: wrap the UMK seed under a passkey-PRF-derived key,
// upload a SINGLE self-contained `wrappedUmk` blob to flagshipserver.com.
// Mirrors apps/mobile/ios/Sources/Flagship/Recovery.swift +
// packages/server-daemon/src/recovery/prf.ts.
//
// The Worker's handleUploadWebauthnRecovery treats `wrappedUmk` as one
// opaque base64 blob — it base64-decodes it and hashes the whole thing.
// So the nonce lives INSIDE the blob (nonce ‖ ct ‖ tag), exactly like
// AcmeAccountKey.wrapForEscrow — there is NO separate nonce field on the
// wire.
//
// The Android passkey-PRF ceremony goes through androidx.credentials
// CredentialManager (PublicKeyCredential + clientDataJSON with the
// `prf` extension). The wrap/unwrap math here is provider-agnostic
// so the same code paths work in tests with a synthetic PRF secret.

package com.flagshipserver.app.keystore

import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

object Recovery {
    /** Fixed PRF input salt — `hmac-secret` is keyed by salt, so the
     *  daemon side, iOS, and Android must all agree. Mirrors the
     *  constant in packages/server-daemon/src/recovery/prf.ts. */
    val PRF_SALT: ByteArray = "flagship/recovery/v1".toByteArray()

    /** HKDF salt for the UMK wrap key. Domain-separated from the ACME
     *  escrow salt (AcmeAccountKey.ESCROW_WRAP_SALT) so the same PRF secret
     *  yields independent wrap keys. */
    val WRAP_SALT: ByteArray = "flagship/recovery-wrap/v1".toByteArray()

    /** Wrap [umkSeed] (32 bytes) under a PRF-derived AES-256-GCM key.
     *  Returns ONE self-contained base64 blob = nonce(12) ‖ ct ‖ tag(16),
     *  ready to ship verbatim as `wrappedUmk`. The nonce is inside the blob
     *  — there is no separate nonce field on the wire. */
    fun wrap(umkSeed: ByteArray, prfSecret: ByteArray): String {
        val key = hkdfSha256(ikm = prfSecret, salt = WRAP_SALT, info = ByteArray(0), length = 32)
        val nonce = ByteArray(12).also(rng::nextBytes)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        val ctWithTag = cipher.doFinal(umkSeed) // ciphertext ‖ 16-byte GCM tag
        val blob = ByteArray(nonce.size + ctWithTag.size)
        System.arraycopy(nonce, 0, blob, 0, nonce.size)
        System.arraycopy(ctWithTag, 0, blob, nonce.size, ctWithTag.size)
        return java.util.Base64.getEncoder().encodeToString(blob)
    }

    /** Reverse of [wrap]: take the single self-contained [wrappedUmkBase64]
     *  blob (nonce ‖ ct ‖ tag) and return the recovered UMK seed bytes.
     *  Throws AEADBadTagException on a wrong PRF secret. */
    fun unwrap(wrappedUmkBase64: String, prfSecret: ByteArray): ByteArray {
        val blob = java.util.Base64.getDecoder().decode(wrappedUmkBase64)
        require(blob.size > 12) { "wrapped UMK blob too short" }
        val nonce = blob.copyOfRange(0, 12)
        val ctWithTag = blob.copyOfRange(12, blob.size)
        val key = hkdfSha256(ikm = prfSecret, salt = WRAP_SALT, info = ByteArray(0), length = 32)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        return cipher.doFinal(ctWithTag)
    }

    private val rng = SecureRandom()

    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        val out = ByteArray(length)
        var t = ByteArray(0)
        var counter = 1
        var written = 0
        while (written < length) {
            mac.reset()
            mac.update(t)
            mac.update(info)
            mac.update(counter.toByte())
            t = mac.doFinal()
            val n = minOf(t.size, length - written)
            System.arraycopy(t, 0, out, written, n)
            written += n
            counter++
        }
        return out
    }
}
